api/grpcserver/http_server.go

```diff
@@ -9,31 +9,29 @@ import (
 	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
 	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
 
-	"github.com/spacemeshos/go-spacemesh/cmd"
 	"github.com/spacemeshos/go-spacemesh/log"
 )
 
 // JSONHTTPServer is a JSON http server providing the Spacemesh API.
 // It is implemented using a grpc-gateway. See https://github.com/grpc-ecosystem/grpc-gateway .
 type JSONHTTPServer struct {
+	ctx    context.Context
 	mu     sync.RWMutex
 	port   int
 	server *http.Server
 }
 
 // NewJSONHTTPServer creates a new json http server.
-func NewJSONHTTPServer(port int) *JSONHTTPServer {
-	return &JSONHTTPServer{port: port}
+func NewJSONHTTPServer(ctx context.Context, port int) *JSONHTTPServer {
+	return &JSONHTTPServer{ctx: ctx, port: port}
 }
 
 // Close stops the server.
 func (s *JSONHTTPServer) Close() error {
 	log.Debug("stopping new json-http service...")
 	server := s.getServer()
 	if server != nil {
-		ctx := cmd.Ctx()
-
-		if err := server.Shutdown(ctx); err != nil {
+		if err := server.Shutdown(s.ctx); err != nil {
```

Comment:
The forwarded context is usually the context of the current call stack. I think it should not be saved in `JSONHTTPServer` and reused here.

It'd require a slightly more refactor:
- accept `ctx context.Context` in `JSONHTTPServer::Close`, potentially renaming to `Shutdown` as it no longer implements `io::Closer` interface,
- accept `ctx context.Context` in `App::Cleanup` and `App::StopServices`,
- remove `ctx context.Context` from `JSONHTTPServer struct`.

💡 the `server.Shutdown` will not wait for the ongoing connections to be closed if the passed context expires.

What do you think?

Reply:
i think it's great. done as suggested. thank you.